src/Microsoft.Diagnostics.Runtime/src/DataTargets/DataTarget.cs

```diff
@@ -205,122 +208,163 @@ public IEnumerable<ModuleInfo> EnumerateModules()
+                        format = DumpFileFormat.CompressedArchive;
+                }
+
+                stream.Position = 0;
+                var result = (stream, format);
+                stream = null;
+                return result;
+            }
+            finally
+            {
+                stream?.Dispose();
+            }
         }
 
         /// <summary>
-        /// Attaches to a live process.
+        /// Attaches to a running process.  Note that if <paramref name="suspend"/> is set to false the user
+        /// of ClrMD is still responsible for suspending the process itself.  ClrMD does NOT support inspecting
+        /// a running process and will produce undefined behavior when attempting to do so.
         /// </summary>
-        /// <param name="processId">The ID of the process to suspend and attach to.</param>
-        /// <returns>A DataTarget instance.</returns>
-        /// <exception cref="ArgumentException">
-        /// The process specified by <paramref name="processId"/> is not running.
-        /// </exception>
-        public static DataTarget SuspendAndAttachToProcess(int processId)
+        /// <param name="processId">The ID of the process to attach to.</param> 
+        /// <param name="suspend">Whether or not to suspend the process.</param>
+        /// <returns>A <see cref="DataTarget"/> instance.</returns>
+        public static DataTarget AttachToProcess(int processId, bool suspend)
         {
             if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
             {
-                LinuxLiveDataReader reader = new LinuxLiveDataReader(processId, suspend: true);
-                return new DataTarget(reader)
-                {
-                    BinaryLocator = new LinuxDefaultSymbolLocator(reader.GetModulesFullPath())
-                };
+                CustomDataTarget customTarget = new CustomDataTarget(new LinuxLiveDataReader(processId, suspend: suspend));
+                return new DataTarget(customTarget);
+            }
+
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                WindowsProcessDataReaderMode mode = suspend ? WindowsProcessDataReaderMode.Suspend : WindowsProcessDataReaderMode.Passive;
+                CustomDataTarget customTarget = new CustomDataTarget(new WindowsProcessDataReader(processId, mode));
+                return new DataTarget(customTarget);
             }
 
-            return new DataTarget(new DbgEngDataReader(processId, invasive: false, 5000));
+            throw new PlatformNotSupportedException(GetPlatformMessage(nameof(AttachToProcess), RuntimeInformation.OSDescription));
         }
 
         /// <summary>
-        /// Attaches to a snapshot process (see https://docs.microsoft.com/windows/win32/api/_proc_snap/).
-        /// This method is only supported on Windows.
+        /// Creates a snapshot of a running process and attaches to it.  This method will pause a running process
```

Comment:
> This method will pause a running process

Really? Why? As I understand this used to be the only safe way to inspect a running process without suspending it.

Reply:
I don't think I finished that comment.  It's supposed to be:  "This method will temporarily pause a running process to generate a snapshot, then resume the process and attach to the snapshot."

The design is:

1.  Use AttachToProcess(pid, suspend: true) - attach "non-invasively", the process is suspended by ClrMD
2.  Use AttachToProcess(pid, suspend: false) - Open process to read memory, this method is only supported if the user suspends the process themselves.
3.  Use CreateSnapshotAndAttach - Creates a "snapshot" and debugs that, fully supported.

For CreateSnapshotAndAttach:  On Windows we have PssCaptureSnapshot which is super fast, on Linux we'll create a coredump to the tmp folder and load it...which will take a lot longer than on Windows, but at least it's a consistent state to debug.